rfc/1216-resolve-cloud-tokens.md

```diff
@@ -0,0 +1,845 @@
+
+```yaml
+apiVersion: batch/v1
+kind: CronJob
+metadata:
+  name: manifest-cronjob-c86481e8
+spec:
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+            cdk8s.io/metadata.addr: Manifest-CronJob-c89809bc
+        spec:
+          containers:
+            - env:
+                - name: BUCKET_NAME
+                  value: my-bucket
+              image: job
+```
+
+However, if the bucket is not explicitly specified (as recommended by the 
+[AWS CDK best practices guide](https://docs.aws.amazon.com/cdk/v2/guide/best-practices.html)):
+
+```ts
+import * as aws from 'aws-cdk-lib';
+import * as k8s from 'cdk8s';
+import * as kplus from 'cdk8s-plus-26';
+
+const k8sApp = new k8s.App();
+const awsApp = new aws.App();
+
+const stack = new aws.Stack(awsApp, 'Stack');
+const manifest = new k8s.Chart(k8sApp, 'Manifest');
+
+// define the bucket without an explicit name
+new aws.aws_s3.Bucket(stack, 'Bucket');
+
+// pass the bucket name to the CronJob container
+new kplus.CronJob(manifest, 'CronJob', {
+ containers: [{ 
+   image: 'job',
+   envVariables: {
+     BUCKET_NAME: kplus.EnvValue.fromValue(bucket.bucketName),
+   }
+ }]
+});
+
+k8sApp.synth();
+awsApp.synth();
+```
+
+The synthesized Kubernetes manifest will not be deployable, as it will not 
+contain the actual value of the bucket name, but rather a string 
+representation of the AWS CDK token, *representing* the bucket name.
+
+```yaml
+apiVersion: batch/v1
+kind: CronJob
+metadata:
+  name: manifest-cronjob-c86481e8
+spec:
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+            cdk8s.io/metadata.addr: Manifest-CronJob-c89809bc
+        spec:
+          containers:
+            - env:
+                - name: BUCKET_NAME
+                  value: ${Token[TOKEN.603]} # whoops, no good
+              image: job
+```
+
+To generate a deployable manifest, an additional lookup phase is required that will
+fetch the value of the bucket name directly from AWS. Currently, customers are forced 
+to implement this lookup by themselves, which presents non trivial challenges and burden. 
+
+This feature aims to alleviate that, and make it as easy as possible to author such 
+complex applications. 
+
+### Why should we _not_ do this?
+
+There are several mechanisms customers could implement this capability without explicit 
+support from the cdk8s team. Given that, one could argue that the effort would not be 
+worth the value. Following is an outline of such mechanisms:
+
+#### Explicit Token Resolution
+
+Instead of passing tokens directly to cdk8s, customers could explicitly fetch its 
+concrete value with a `fetchToken` function they implement:
+
+```ts
+...
+
+// define the bucket without an explicit name
+new aws.aws_s3.Bucket(stack, 'Bucket');
+
+// fetch its name directly from AWS
+const bucketName = fetchToken(bucket.bucketName);
```

Comment:
Unless Pulumi, this wouldn't work, would it? Because we'd need a deploy first as well?